### PR TITLE
feature(environment): Allows plugins to detect the site environment

### LINKED ISSUE #8016: Introduce elgg_development_mode
elgg_maintenance_mode was a great addition, but I am missing a way to differentiate between development and production modes. 'debug' config doesn't really cut it.

I think we need a global flag that plugins could use to load appropriate settings, and do things differently when in development mode. This would also help with #7946, we could just only show deprecation warning when in development mode.

### PR DESCRIPTION
**Pull request by mrclay**
Adds an “env” service to detect name/production status
Allows this to be set via settings.php from an array (or factory function)
Shows the environment values on the server info page
Env name shown in admin header (if name !== default)
Adds TOC to services docs
Replaces some $CONFIG keys with class constants
Links the user name in admin header

Fixes #8016

### CODE REVIEW

**Comment by mrclay** [comment]
TL;DR this gives us by default:
```php
elgg()->env->getName(); // "default"
elgg()->env->isProd(); // true
```
And configurable via settings.php

File: docs/guides/services.rst [hunk 1]
```diff
@@ -21,11 +28,36 @@ deprecated and will be supported at least through Elgg 2.x. If they will reduce
...
 
 Service: config
----------------
+===============
 
-An instance of ``Elgg\Services\ConfigInterface``, this is for getting and setting various system
+An instance of ``Elgg\Services\Config``, this is for getting and setting various system
 configuration values.
 
+Service: env
```
Comment (hypeJunction) [inline docs/guides/services.rst]: Perhaps we should mark config and env as code?

**Comment by hypeJunction** [comment]
The code LGTM. Does this need manual testing?

**Comment by mrclay** [comment]
Freshly rebased

File: docs/guides/services.rst (hunk 1 above)
Comment (ewinslow) [inline docs/guides/services.rst]: Agreed

**Comment by ewinslow** [comment]
Why separate `isProd()` method/concept distinct from "name"? I was figuring one of the names would just be "prod".

**Comment by mrclay** [comment]
Site could have several production instances. I want an easy way to tell if I'm production without knowledge of all their names.

**Comment by ewinslow** [comment]
This is very different from how I think of "environment" which is not "server instance" but "prod" "staging" "local". CF laravel...

**Comment by mrclay** [comment]
It boils down to whether comparing the env name to "prod" is good enough for most read purposes? Certainly plugins won't being able to guess any other names. Do we want to suggest particular names? "prod", "staging", "local", "ci" ?

**Comment by ewinslow** [comment]
Yea, might not be a bad idea to suggest some. I like those 4.

**Comment by ewinslow** [comment]
Maybe "test" instead of "ci". "ci" is a wee bit more cryptic. No strong opinion either way.
